src/MemoryStream.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+    Copyright 2016-2019 Arisotura
+
+    This file is part of melonDS.
+
+    melonDS is free software: you can redistribute it and/or modify it under
+    the terms of the GNU General Public License as published by the Free
+    Software Foundation, either version 3 of the License, or (at your option)
+    any later version.
+
+    melonDS is distributed in the hope that it will be useful, but WITHOUT ANY
+    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+    FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with melonDS. If not, see http://www.gnu.org/licenses/.
+*/
+
+#include "MemoryStream.h"
+
+MemoryStream::MemoryStream()
+{
+	// Create new array of chunks, and an initial chunk.
+	chunks = new u8*[MAX_CHUNKS];
+	memset(chunks, 0, MAX_CHUNKS * sizeof(u8*));
+	chunks[0] = new u8[CHUNK_SIZE];
+	
+	pos = 0;
+	size = 0;
+	firstNonsequentialChunk = 0;
+}
+MemoryStream::MemoryStream(u8* data, s32 len)
+{
+	chunks = NULL;
+	Init(data, len);
+}
+MemoryStream::~MemoryStream()
+{
+	DeInit();
+}
+
+// Make chunks point to data; does not copy data.
+void MemoryStream::Init(u8* data, s32 len)
+{
+	if (chunks) DeInit();
+
+	if (len > CHUNK_SIZE * MAX_CHUNKS)
+		throw "data array too big";
```

Comment:
this would introduce exceptions into the code base, maybe we can come up with something different